src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/SystemModulesPlugin.java

```diff
@@ -587,440 +573,406 @@ private void dedups(ModuleDescriptor md) {
-                                             + ClassWriter.COMPUTE_FRAMES);
-            cw.visit(Opcodes.V1_8,
-                     ACC_FINAL+ACC_SUPER,
-                     className,
-                     null,
-                     "java/lang/Object",
-                     new String[] { "jdk/internal/module/SystemModules" });
-
-            // generate <init>
-            genConstructor(cw);
+        public byte[] getClassWriter(Configuration cf) {
+            return Classfile.build(classDesc,
+                    clb -> {
+                        clb.withFlags(ACC_FINAL + ACC_SUPER);
+                        clb.withInterfaceSymbols(List.of(CD_SYSTEM_MODULES));
+                        clb.withVersion(52, 0);
 
-            // generate hasSplitPackages
-            genHasSplitPackages(cw);
+                        // generate <init>
+                        genConstructor(clb);
 
-            // generate hasIncubatorModules
-            genIncubatorModules(cw);
+                        // generate hasSplitPackages
+                        genHasSplitPackages(clb);
 
-            // generate moduleDescriptors
-            genModuleDescriptorsMethod(cw);
+                        // generate hasIncubatorModules
+                        genIncubatorModules(clb);
 
-            // generate moduleTargets
-            genModuleTargetsMethod(cw);
+                        // generate moduleDescriptors
+                        genModuleDescriptorsMethod(clb);
 
-            // generate moduleHashes
-            genModuleHashesMethod(cw);
+                        // generate moduleTargets
+                        genModuleTargetsMethod(clb);
 
-            // generate moduleResolutions
-            genModuleResolutionsMethod(cw);
+                        // generate moduleHashes
+                        genModuleHashesMethod(clb);
 
-            // generate moduleReads
-            genModuleReads(cw, cf);
+                        // generate moduleResolutions
+                        genModuleResolutionsMethod(clb);
 
-            return cw;
+                        // generate moduleReads
+                        genModuleReads(clb, cf);
+                    });
         }
 
         /**
          * Generate bytecode for no-arg constructor
          */
-        private void genConstructor(ClassWriter cw) {
-            MethodVisitor mv = cw.visitMethod(ACC_PUBLIC, "<init>", "()V", null, null);
-            mv.visitVarInsn(ALOAD, 0);
-            mv.visitMethodInsn(INVOKESPECIAL,
-                               "java/lang/Object",
-                               "<init>",
-                               "()V",
-                               false);
-            mv.visitInsn(RETURN);
-            mv.visitMaxs(0, 0);
-            mv.visitEnd();
+        private void genConstructor(ClassBuilder clb) {
+            clb.withMethod("<init>", MethodTypeDesc.of(CD_void),
+                    ACC_PUBLIC, mb -> mb.withFlags(ACC_PUBLIC).withCode( cob -> {
```

Comment:
Why `withFlags(ACC_PUBLIC)` when the flags is already given in `withMethod`'s 3rd argument? Same for occurrences below.

Reply:
It has been redundant, now more effective `withMethodBody` method is used.

Reply:
I also had the same comment in PR [12945](https://github.com/openjdk/jdk/pull/12945).   Similar change should apply to PR 12945.